8234 persist attorney re write info

youngfreezyVA opened the pull request:
Resolves #8234 and #8434 

### Description
this issue describes it best:
https://github.com/department-of-veterans-affairs/caseflow/issues/8234
this pr only handles AMA appeals

### Acceptance Criteria
- [ ] in the `/submit` view, if checking out as an attorney, and this appeal has been returned to you by a judge, the following fields should be pre populated: 

- the judges instructions `AttorneyRewriteTask.instructions`
- the document id -
- the reviewing judge 
- overtime (?)

### Testing Plan
1. as a judge with cases, select an AMA appeal, return it to an attorney (remember the note)
2. switch users to that attorney, checkout
3. once you're at submit, all fields should be pre populated

@aroltsch @lowellrex please let me know if the naming of the functions/variables is accurate/explicit enough.

youngfreezyVA commented on spec/feature/queue/ama_queue_spec.rb:
```diff
@@ -276,27 +276,6 @@ def valid_document_id
         expect(page).to have_content("Task assigned to #{other_organization.name}")
         expect(Task.last.instructions.first).to eq(instructions)
       end
-
```
I didn't delete this test. no idea how this got in the diff. I can add it back.

aroltsch replied:
Yes, please.

aroltsch commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,13 +152,22 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
+    {
+      document_id: latest_attorney_case_review&.document_id,
+      assigned_judge: object.assigned_judge,
+      note: object.attorney_rewrite_task_for_appeal&.instructions&.first,
```
So in this case assigned_judge should be: 
```
object.attorney_rewrite_task_for_appeal&.assigned_by&.full_name
```

aroltsch replied:
I think `note` and `instructions` are two separate things and we should display them both. `note` is what attorney entered when going through the checkout and `instructions` is what a judge entered.  We need to ask our design team on where they want instructions to be displayed.

aroltsch commented on app/models/appeal.rb:
```diff
@@ -137,6 +137,10 @@ def reviewing_judge_name
   end
 
+  def attorney_rewrite_task_for_appeal
+    tasks.find { |t| t.is_a?(AttorneyRewriteTask) && (t.appeal_id == id) }
```
The second part in this find statement is redundant. 

when you call `tasks` inside the appeal class, it returns all tasks where t.appeal_id == id because of an appeal `has_many :tasks` association 

Also you probably want to look for non-completed AttorneyRewriteTask tasks because there are could be multiple.

aroltsch commented:
Hey Fareez, nice start! I've added a few comments. Also, we should add some unit tests for the methods you added on the BE.

youngfreezyVA commented:
> Hey Fareez, nice start! I've added a few comments. Also, we should add some unit tests for the methods you added on the BE.

Hey @aroltsch addressed all your comments, please have a look! Thanks.

aroltsch commented on app/models/appeal.rb:
```diff
@@ -137,6 +137,10 @@ def reviewing_judge_name
   end
 
+  def attorney_rewrite_task_for_appeal
+    tasks.find { |t| t.is_a?(AttorneyRewriteTask) && (t.status == Constants.TASK_STATUSES.in_progress) }
```
I think more accurate way would be to also include assigned statuses. Also, you could do the filtering in SQL query though ActiveRecord instead of doing it in Ruby.

```
tasks.where(type: "AttorneyRewriteTask", status: [Constants.TASK_STATUSES.in_progress, Constants.TASK_STATUSES.assigned]
```

aroltsch commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,13 +152,23 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
-
+  # this should probably be moved to `appeal.rb` but didn't want to break anything
   def latest_attorney_case_review
     AttorneyCaseReview.where(task_id: Task.where(appeal: object).pluck(:id)).order(:created_at).last
```
Let's set it to an instance variable and use a conditional assignment operator here so it won't call this query every time you invoke this method. 

```
def latest_attorney_case_review
    @latest_attorney_case_review ||= AttorneyCaseReview.where(task_id: Task.where(appeal: object).pluck(:id)).order(:created_at).last
```

Let's do the same for `attorney_rewrite_task_for_appeal` method.

aroltsch commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -175,25 +186,33 @@ class SubmitDecisionView extends React.PureComponent<Props> {
         name="notes"
-        value={decisionOpts.note}
+        // the judges instructions take precedence in the note 
+        // over the note that the attorney previously wrote in checkout
```
Is that what design team suggested? Seems a bit weird to me

youngfreezyVA replied:
nope! that was my own solution - I didn't speak with them, sorry! let me know what we should do here.

aroltsch replied:
I think we should ask our design team (LP, Sneha, Allyce). Could you post this question on slack to them. I personally think we should put judge's instructions somewhere else but we might need design for that and could do it as a separate ticket.

youngfreezyVA replied:
ok, i will remove the logic about the judges rewrite instructions from this PR - resolving this convo since that will be handled in another itcket

aroltsch commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -133,15 +133,26 @@ class SubmitDecisionView extends React.PureComponent<Props> {
 
+  getDefaultJudgeSelector = () => {
+    const legacyJudgeSelector = _.get(this.props, 'task.addedByCssId', null);
+    const amaJudgeSelector = _.get(this.props.appeal, 'assignedJudge.full_name');
```
If `assignedJudge` is nil would it still work?

youngfreezyVA replied:
yes, the `_.get` method is basically a null safe operator for javascript, it resolves to undefined by default, it's the same thing as `&` in Rails

aroltsch commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,13 +152,23 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
     latest_attorney_case_review&.id
   end
 
+  attribute :attorney_case_review_details do
```
I would name it `attorney_case_rewrite_details` to differentiate between those two. Although I don't feel strongly about it.

aroltsch commented on app/models/appeal.rb:
```diff
@@ -152,6 +152,12 @@ def reviewing_judge_name
+  def attorney_rewrite_task_for_appeal
+    @attorney_rewrite_task_for_appeal ||=
+      tasks.where(type: "AttorneyRewriteTask",
+                  status: [Constants.TASK_STATUSES.in_progress, Constants.TASK_STATUSES.assigned])
```
let's also sort it by created_at to get the latest one (you can probably use `max_by` method for that)

aroltsch replied:
I think we should create a new method on task similar to this: https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/task.rb#L39

We can call it `open_statuses` or something. @lowellrex do you have thoughts?

aroltsch commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,14 +152,24 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
+  attribute :attorney_case_rewrite_details do
+    {
+      document_id: latest_attorney_case_review&.document_id,
+      assigned_judge: object.attorney_rewrite_task_for_appeal&.first&.assigned_by,
```
If you use `max_by` above it will return on element instead of a collection.

`where` returns a collection

lowellrex replied:
I don't think we need to return this value in the serializer because it is already available to us in `SubmitDecisionView` by calling `this.props.task.assignedBy`.

aroltsch commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -116,7 +120,14 @@ class SubmitDecisionView extends React.PureComponent<Props> {
 
     const issuesToPass = !isLegacyAppeal && amaDecisionIssues ? decisionIssues : issues;
-    const payload = buildCaseReviewPayload(checkoutFlow, decision, userRole, issuesToPass, { isLegacyAppeal });
+    // not sure why this is happening, it's causing tests to fail
```
I don't understand what this comment relates to

youngfreezyVA replied:
if the frontend sends decision.opts.overtime as null to the backend, a 400 error is returned saying 'overtime is not on the list' -- this only happens on the automated tests though, i wasnt able to reproduce it locally. i think this is a backend bug. to work around this I had to make sure it was false instead of null

aroltsch commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,14 +152,24 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
 
+  attribute :attorney_case_rewrite_details do
+    {
+      document_id: latest_attorney_case_review&.document_id,
```
Is it necessary to include it here? There is already a document_id included: 
https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/serializers/work_queue/appeal_serializer.rb#L147

lowellrex replied:
Should be able to access it in `SubmitDecisionView` by `this.props.appeal.documentId`.

youngfreezyVA commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -37,46 +36,51 @@ import {
     const missingParams = _.filter(requiredParams, (param) => !_.has(decisionOpts, param) || !decisionOpts[param]);
 
-    return !missingParams.length;
+    const isValid = !missingParams.length;
```
initially for debugging and then I figured it's also more explicit

lowellrex replied:
Can `isValid` be false?

lowellrex commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,14 +152,24 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
+    {
+      document_id: latest_attorney_case_review&.document_id,
+      assigned_judge: object.attorney_rewrite_task_for_appeal&.first&.assigned_by,
+      note_from_judge: object.attorney_rewrite_task_for_appeal&.first&.instructions&.first,
```
We usually keep most task-related information and workflow-related information (like attorney and judge checkout) separate from one another. We currently display notes from the judge back to the attorney in the task snapshot area:

![image](https://user-images.githubusercontent.com/32683958/52867107-2b767f00-310e-11e9-9787-9da93f267e76.png)

Also, it looks like we don't actually end up using `note_from_judge` anywhere on the frontend. Can we get rid of it?

youngfreezyVA replied:
yes! I was planning on doing that for this PR, as the UX team is still deciding if/where we need to place that somewhere (will be a separate task)

lowellrex commented on app/models/appeal.rb:
```diff
@@ -152,6 +152,12 @@ def reviewing_judge_name
 
+  def attorney_rewrite_task_for_appeal
+    @attorney_rewrite_task_for_appeal ||=
+      tasks.where(type: "AttorneyRewriteTask",
```
Let's use `AttorneyRewriteTask.name` here so that the static analyzer can help us catch any issues if the name of this class changes somewhere down the line.

lowellrex commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,14 +152,24 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
   def latest_attorney_case_review
-    AttorneyCaseReview.where(task_id: Task.where(appeal: object).pluck(:id)).order(:created_at).last
+    @latest_attorney_case_review ||=
+      AttorneyCaseReview.where(task_id: Task.where(appeal: object).pluck(:id)).order(:created_at).last
```
Can we take advantage of the [existing `Appeal.attorney_case_reviews` method](https://github.com/department-of-veterans-affairs/caseflow/blob/fa41c7799926426996592a0b8cacc2a7542be0e5/app/models/appeal.rb#L133) here?

youngfreezyVA replied:
based on the fact that we don't need alot the information I thought we did, I can just grab all the info I need from this method: ```def latest_attorney_case_review
    @latest_attorney_case_review ||=
      AttorneyCaseReview.where(task_id: Task.where(appeal: object).pluck(:id)).order(:created_at).last
  end```

youngfreezyVA replied:
oh I may have misunderstood your question - this isn't a new method I created, it was already there, I just modified it per @aroltsch feedback so that it only makes the query once (initially) when that method is called in the class

lowellrex commented on client/app/queue/JudgeSelectComponent.jsx:
```diff
@@ -48,7 +48,11 @@ class JudgeSelectComponent extends React.PureComponent<Props> {
+       // for legacy
+       _.find(judges, { css_id: this.props.judgeSelector }) ||
+       // for ama
+       _.find(judges, { full_name: this.props.judgeSelector });
```
What is this change?

lowellrex replied:
Ahhh I see that we've [changed the interface](https://github.com/department-of-veterans-affairs/caseflow/pull/9267/files#diff-dc6af3991f74c682c598e9a98c97bf8dR147) for this component.

I think we can use the ID of the judge instead of the name so we guarantee the ID will be unique when the name might not be. This also allows us to not have to worry about changing formatting of names as they appear in this list.

![image](https://user-images.githubusercontent.com/32683958/52869381-878fd200-3113-11e9-88c1-df1b4afe596f.png)

lowellrex commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -133,6 +144,13 @@ class SubmitDecisionView extends React.PureComponent<Props> {
   };
 
+  getDefaultJudgeSelector = () => {
+    const legacyJudgeSelector = _.get(this.props, 'task.addedByCssId', null);
```
I think we can use the existing `isLegacy` attribute of the `this.props.task` property to determine whether to return a CSS ID or the User ID. Maybe something like:

```jsx
getDefaultJudgeSelector = () => this.props.task.isLegacy ? this.props.task.addedByCssId : this.props.task.assignedBy.pgId;
```

youngfreezyVA replied:
awesome, thanks!

lowellrex commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -37,46 +36,51 @@ import {
+  // and it comes from this.props.appeal.attorneyCaseRewriteDetails instead
+  // if you don't do this you will get validation errors and a 400 for a missing document_id
+  // this is also needed to keep the onChange values in sync
+  extendDecisionOptsWithAttorneyCheckOutInfo = (decision, attorneyCaseRewriteDetails) => {
```
Can we name this function something along the lines of `setInitialDecisionOptions`?

lowellrex commented on client/app/queue/SubmitDecisionView.jsx:
```diff
@@ -116,7 +120,14 @@ class SubmitDecisionView extends React.PureComponent<Props> {
+    // the backend can't handle a null response at this point
+
+    if (decision.opts.overtime === null) {
+      decision.opts.overtime = false;
```
Let's set this value to false when [we pluck out the initial values](https://github.com/department-of-veterans-affairs/caseflow/pull/9267/files#diff-dc6af3991f74c682c598e9a98c97bf8dR58) of the other fields.

lowellrex commented on app/models/serializers/work_queue/appeal_serializer.rb:
```diff
@@ -152,14 +152,24 @@ class WorkQueue::AppealSerializer < ActiveModel::Serializer
+      assigned_judge: object.attorney_rewrite_task_for_appeal&.first&.assigned_by,
+      note_from_judge: object.attorney_rewrite_task_for_appeal&.first&.instructions&.first,
+      overtime: latest_attorney_case_review&.overtime,
+      note_from_attorney: latest_attorney_case_review&.note
```
I think `overtime` and `note_from_attorney` are the only two new fields we need to return in the appeal serializer for this PR.

aroltsch commented on spec/factories/task.rb:
```diff
@@ -105,6 +105,11 @@
       parent { create(:ama_judge_task) }
     end
 
+    factory :ama_attorney_rewrite_task do
```
@youngfreezy I can see you added this factory but I am not seeing a feature test for that. Could you please add the test.
